### PR TITLE
Added coremltools

### PR DESCRIPTION
**Pull request by TomAugspurger**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

File: recipes/coremltools/meta.yaml
Comment (TomAugspurger) [inline recipes/coremltools/meta.yaml:70]: @1duo and @aseem could you comment here to confirm that you're both OK with being maintainers here?

File: recipes/coremltools/build.sh
Comment (TomAugspurger) [inline recipes/coremltools/build.sh:11]: I think the linux64 build failure at https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=225770&view=logs&j=10bda42a-d420-5c28-a43f-bc4e66405e5c&t=bef775d0-a1a2-55a3-9721-17147029baf0 is is from this.

```
/home/conda/staged-recipes/build_artifacts/coremltools_1602772503997/work/deps/pybind11/include/pybind11/common.h:72:10: fatal error: Python.h: No such file or directory
 #include <Python.h>
          ^~~~~~~~~~
```

Does conda or conda-forge have a standard way of getting the this path (i.e. whether it's `3.8` vs `3.8m`)?

**Comment by TomAugspurger** [comment]
The MacOS build at https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=225770&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310 failed with

```
Scanning dependencies of target coremlpython
[ 85%] Building CXX object CMakeFiles/coremlpython.dir/coremlpython/CoreMLPython.mm.o
/Users/runner/miniforge3/conda-bld/coremltools_1602772502563/work/coremlpython/CoreMLPython.mm:1:9: fatal error: 'CoreML/CoreML.h' file not found
#import <CoreML/CoreML.h>
        ^~~~~~~~~~~~~~~~~
```

I suppose that's because coremltools requires MacOS>=10.13 to build? https://github.com/apple/coremltools/blob/523d5e03d86c26267ee6bdf17dd20f6ce6bdadd7/scripts/build.sh#L80-L82

So I'll follow https://conda-forge.org/docs/maintainer/knowledge_base.html#requiring-newer-macos-sdks to require 10.13.

File: recipes/coremltools/build.sh
Comment (chrisburr) [inline recipes/coremltools/build.sh:9]: If this is actually needed, see: https://conda-forge.org/docs/maintainer/knowledge_base.html#requiring-newer-macos-sdks

(I don't think it will do anything in staged recipes but it will in the feedstock)

File: recipes/coremltools/meta.yaml
Comment (chrisburr) [inline recipes/coremltools/meta.yaml:26]:
```suggestion
```

Comment (xhochy) [inline recipes/coremltools/meta.yaml:26]: This should only be in `host`

File: recipes/coremltools/meta.yaml
Comment (xhochy) [inline recipes/coremltools/meta.yaml:29]: Can you omit the constraint so it uses the `numpy` from the pinnings file (1.16)?

File: recipes/coremltools/meta.yaml
Comment (xhochy) [inline recipes/coremltools/meta.yaml:33]:
```suggestion
    - {{ pin_compatible('numpy') }}
```

File: recipes/coremltools/meta.yaml
Comment (chrisburr) [inline recipes/coremltools/meta.yaml:29]: If this is using the numpy C API, see: https://conda-forge.org/docs/maintainer/knowledge_base.html#building-against-numpy

```suggestion
    - numpy
```

If it isn't, this can probably be removed from `host`.

File: recipes/coremltools/build.sh
Comment (xhochy) [inline recipes/coremltools/build.sh:9]: Doesn't this work with 10.9?

File: recipes/coremltools/meta.yaml
Comment (chrisburr) [inline recipes/coremltools/meta.yaml:41]:
```suggestion
```

File: recipes/coremltools/build.sh
Comment (xhochy) [reply recipes/coremltools/build.sh:11]: This should get you the correct path, not sure whether there is a simpler solution. Normally passing one of `PYTHON_EXECUTABLE`,  `Python_EXECUTABLE`, `Python3_EXECUTABLE` is enough. Note that you may need a different casing based on the CMake version.
```suggestion
    -DPYTHON_INCLUDE_DIR=$(${PYTHON} -c 'import sysconfig; print(sysconfig.get_paths()["include"])') \
```

File: recipes/coremltools/meta.yaml
Comment (TomAugspurger) [reply recipes/coremltools/meta.yaml:29]: Thanks. I believe it does use the NumPy C API: https://github.com/apple/coremltools/blob/523d5e03d86c26267ee6bdf17dd20f6ce6bdadd7/CMakeLists.txt#L142-L152

File: recipes/coremltools/build.sh
Comment (TomAugspurger) [reply recipes/coremltools/build.sh:9]: I don't believe so: https://github.com/conda-forge/staged-recipes/pull/12900#issuecomment-709388849. It uses headers that are only available in 10.13+ (I think).

File: recipes/coremltools/meta.yaml
Comment (chrisburr) [reply recipes/coremltools/meta.yaml:29]: Yup it definitely is, so you'll need the pin_compatible stuff 👍

**Comment by TomAugspurger** [comment]
Making progress, the linux build passed!

The change to require MacOS 10.13 didn't seem to take effect in https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=225802&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310. e.g. 

```
+++ echo 'export CONDA_BUILD_SYSROOT='\''/Applications/Xcode_11.7.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk'\'''
+++ echo 'export MACOSX_DEPLOYMENT_TARGET='\''10.9'\'''
+++ [[ -d /Applications/Xcode_11.7.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk ]]
+++ echo 'Found CONDA_BUILD_SYSROOT: /Applications/Xcode_11.7.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk'
Found CONDA_BUILD_SYSROOT: /Applications/Xcode_11.7.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk
```

@chrisburr is this what you meant by "(I don't think it will do anything in staged recipes but it will in the feedstock)" in https://github.com/conda-forge/staged-recipes/pull/12900#discussion_r505620445?

**Comment by chrisburr** [comment]
> @chrisburr is this what you meant by "(I don't think it will do anything in staged recipes but it will in the feedstock)" in #12900 (comment)?

Yup staged recipes isn't very clever when it comes to more advanced things like the SDK version.

It's probably easiest to skip macOS for now and add it again in the feedstock.

File: recipes/coremltools/build.sh
Comment (chrisburr) [inline recipes/coremltools/build.sh:9]: One minor thing and then it looks good once the other maintainers confirm:

```suggestion
    -DCMAKE_OSX_DEPLOYMENT_TARGET=${MACOSX_DEPLOYMENT_TARGET} \
```

Comment (TomAugspurger) [reply recipes/coremltools/build.sh:9]: Good catch!

**Review by ocefpaf — APPROVED** [review]
LGTM @chrisburr and @xhochy do you have any other review comments or is this ready to merge?

**Comment by xhochy** [comment]
Waiting for maintainer approval, otherwise this can go in.

File: recipes/coremltools/meta.yaml
Comment (aseemw) [reply recipes/coremltools/meta.yaml:75]: Yes, confirmed.

Comment (1duo) [reply recipes/coremltools/meta.yaml:75]: Yes.

**Comment by TomAugspurger** [comment]
Thanks all! I'll confirm that the SDK version is honored correctly in the feedstock.

**Pull request merged**
